### PR TITLE
[ci] Support to skip vstest using include/exclude config file.

### PR DESCRIPTION
**Pull request by liushilongbuaa**
#### Why I did it
Some changes didn't change vs image. We should not run vstest in such cases.
#### How I did it
```
example:
├── folderA
│  ├──  fileA (skip vstest)
│  ├──  fileB
│  └──  fileC
If we want to skip vstest when changing /folderA/fileA, and not skip vstest when changing fileB or fileC.

vstest-include:
^folderA/fileA

vstest-exclude:
^folderA
```
#### Which release branch to backport (provide reason below if selected)

- [ ] 201811
- [ ] 201911
- [ ] 202006
- [ ] 202012
- [ ] 202106
- [ ] 202111

### CODE REVIEW

File: azure-pipelines.yml [hunk 1]
```diff
@@ -86,6 +86,7 @@ stages:
   jobs:
   - job:
     pool: sonictest
+    condition: ne(stageDependencies.BuildVS.vs.outputs['SetVar.SKIP_VSTEST'], 'YES')
```
Comment (xumia) [inline azure-pipelines.yml:89]: There is another condition is VSBuild succeeded, right?
What is the behavior, if the VSBuild Failed, but the SKIP_VSTEST is set.

Can we set the condition in Stage level, not for each of the test job?

File: .azure-pipelines/template-skipvstest.yml [hunk 2]
```diff
@@ -0,0 +1,19 @@
+steps:
+- script: |
+    echo "##vso[task.setvariable variable=SKIP_VSTEST;isOutput=true]NO"
+    set -ex
+    [[ "$(Build.Reason)" != "PullRequest" ]] && echo $(Build.Reason) && exit 0
+    for i in $(git diff HEAD~ | grep '^diff ' | awk '{print$NF}' | sed 's#b/##')
+    do
+      # if code is not in /platform, /files/build/versions and .azure-pipelines, don't skip.
+      echo $i | grep -v "^platform" | grep -v "^files/build/versions" | grep -v "^.azure-pipelines" && echo $i && exit 0
```
Comment (xumia) [inline .azure-pipelines/template-skipvstest.yml:9]: Can we make the filter patterns configurable?

File: .azure-pipelines/template-skipvstest.yml [hunk 3]
```diff
@@ -0,0 +1,19 @@
+steps:
+- script: |
+    echo "##vso[task.setvariable variable=SKIP_VSTEST;isOutput=true]NO"
```
Comment (xumia) [inline .azure-pipelines/template-skipvstest.yml:3]: Can we remove the default value?

File: .azure-pipelines/template-skipvstest.yml [hunk 4]
```diff
@@ -0,0 +1,19 @@
+steps:
+- script: |
```
Comment (xumia) [inline .azure-pipelines/template-skipvstest.yml:2]: We can exclude the script if the Build.Reason is not "PullRequest", right?
```
steps:
- ${{ eq(variables["Build.Reason"], "PullRequest") }}:
  - script:
     ....  
 ```

If not work, another option:
```
steps:
- script: |
    ....
  condition: eq($(Build.Reason)", "PullRequest")
```

File: azure-pipelines.yml (hunk 1 above)
Comment (liushilongbuaa) [reply azure-pipelines.yml]: yes, it needs succeeded()
We can't set in stage level. it don't work. https://dev.azure.com/mssonic/build/_build/results?buildId=108401&view=results

File: .azure-pipelines/template-skipvstest.yml [hunk 5]
```diff
@@ -0,0 +1,18 @@
...
+        echo $i | grep -v "^platform" | grep -v "^files/build/versions" | grep -v "^.azure-pipelines" && echo $i && exit 0
+        # if code is in /platform/vs, don't skip
+        echo $i | grep "^platform/vs" && echo $i && exit 0
+        # if code is in .azure-pipelines/run-test-template.yml, don't skip
+        echo $i | grep "^.azure-pipelines/run-test-template.yml" && echo $i && exit 0
+      done
+      set +x
+      echo "Skip vstest jobs"
+      echo "##vso[task.setvariable variable=SKIP_VSTEST;isOutput=true]YES"
+    name: SetVar
```
Comment (xumia) [inline .azure-pipelines/template-skipvstest.yml:17]: How about changing a meaningful name? For example, EvaluatePRChange, or any better name.
And change the display name as well.

File: .azure-pipelines/template-skipvstest.yml [hunk 6]
```diff
@@ -0,0 +1,11 @@
+steps:
+- ${{ if eq(variables['Build.Reason'], 'PullRequest') }}:
+  - script: |
+      set -ex
+      git diff HEAD~ | grep '^diff ' | awk '{print$NF}' | sed 's#b/##' | grep -v -f .azure-pipelines/skip-vstest-include && exit 0
```
Comment (xumia) [inline .azure-pipelines/template-skipvstest.yml:5]: Please change to use
```
git diff --name-only HEAD~1
```

File: .azure-pipelines/template-skipvstest.yml (hunk 5 above)
Comment (liushilongbuaa) [reply .azure-pipelines/template-skipvstest.yml:11]: We can add other vars in this step in future.
We can change display name

**Review by xumia — APPROVED** [review]

**Pull request merged**